src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -17,115 +19,413 @@
+
+        internal void ThrowIfSymbolsChanged()
+        {
+            if (_symbols != null)
+            {
+                var endHash = _symbols.VersionHash;
+                if (_symbolHash != endHash)
+                {
+                    throw new InvalidOperationException($"SymbolTable was mutated during binding of {_expression}");
+                }
+            }
+        }
+
+        public ParseResult ApplyParse()
+        {
+            if (_expression == null)
+            {
+                throw new InvalidOperationException($"Must call {nameof(SetText)} before calling ApplyParse().");
+            }
+
+            if (this.Parse == null)
+            {
+                var result = Engine.Parse(_expression, Source.Config.Features, _parserOptions);
+                this.Parse = result;
+
+                _errors.AddRange(this.Parse.Errors);
+            }
+
+            return this.Parse;
+        }
+
+        internal Formula GetParseFormula()
+        {
+            var parseResult = this.ApplyParse();
+
+            var expression = parseResult.Text;
+            var culture = parseResult.Options.Culture;
+            var formula = new Formula(expression, culture);
+            formula.ApplyParse(parseResult);
+
+            return formula;
+        }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="CheckResult"/> class.
+        /// Call to run binding. 
+        /// This will compute types on each node, enabling calling <see cref="GetNodeType(TexlNode)"/>.
         /// </summary>
-        public CheckResult()
+        public void ApplyBinding()
         {
+            ApplyBindingInternal();
         }
 
-        internal CheckResult(ParseResult parse, TexlBinding binding = null) 
-            : this(parse, null, binding)
+        // Apply and return the binding. 
+        internal TexlBinding ApplyBindingInternal()
         {
+            if (_binding == null)
+            {
+                if (!this._setBindingCalled)
+                {
+                    throw new InvalidOperationException($"Must call {nameof(SetBindingInfo)} before calling {nameof(ApplyBinding)}.");
+                }
+
+                (var binding, var combinedSymbols) = Source.ComputeBinding(this);
+
+                this.ThrowIfSymbolsChanged();
+                
+                // Don't modify any fields until after we've verified the symbols haven't change.
+
+                this._binding = binding;
+                this.AllSymbols = combinedSymbols;
+
+                // Add the errors
+                var bindingErrors = ExpressionError.New(binding.ErrorContainer.GetErrors(), CultureInfo);
+                _errors.AddRange(bindingErrors);
+
+                if (this.IsSuccess && !this.HasDeferredArgsWarning)
+                {
+                    // TODO: Fix FormulaType.Build to not throw exceptions for Enum types then remove this check
+                    if (binding.ResultType.Kind != DKind.Enum)

```

Comment:
Can we remove this check now?

Reply:
I'm not sure - this is just existing code being moved; and I'm trying to preserve behavior with the refactor.